chapter_3/hello_lib/CMakeLists.txt

```diff
@@ -30,23 +25,45 @@ add_library(hello SHARED)
 # project version * SOVERSION set the compatibility  version for the library to
 # the major number of the version
 set_target_properties(
-  hello
-  PROPERTIES DEBUG_POSTFIX ${CMAKE_DEBUG_POSTFIX}
-             VERSION ${PROJECT_VERSION}
-             SOVERSION ${PROJECT_VERSION_MAJOR})
+    hello
+    PROPERTIES DEBUG_POSTFIX ${CMAKE_DEBUG_POSTFIX}
+               VERSION ${PROJECT_VERSION}
+               SOVERSION ${PROJECT_VERSION_MAJOR}
+)
 
 # add sources to the library target
 target_sources(
-  hello
-  PRIVATE src/hello.cpp src/internal.cpp
-  PUBLIC include/hello/hello.hpp)
+    hello
+    PRIVATE src/hello.cpp src/internal.cpp
+    PUBLIC include/hello/hello.hpp
+)
 
 # define the C++ standard needed to compile this library and make it visible to
 # dependers
 target_compile_features(hello PUBLIC cxx_std_17)
 
 # set the include directories
 target_include_directories(
-  hello
-  PRIVATE src/hello
-  PUBLIC include)
+    hello
+    PRIVATE src/hello
+    PUBLIC include
+)
+
+# An alternative way to create an export header Use either
+# CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS or generate_export_header
+if(MSVC)
+    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS TRUE)
+    set(BUILD_SHARED_LIBS TRUE)
+endif()
+
+# if using limited visibility, set CXX_VISIBILILTY_PRESET to "hidden"
+include(GenerateExportHeader)
+set_property(TARGET hello PROPERTY CXX_VISIBILITY_PRESET "hidden")
+# Hide inline functions by default
+set_property(TARGET hello PROPERTY CMAKE_VISIBILITY_INLINES_HIDDEN TRUE)
```

Comment:
`CMAKE_VISIBILITY_INLINES_HIDDEN` is the name of the variable, not the property ([VISIBILITY_INLINES_HIDDEN](https://cmake.org/cmake/help/latest/prop_tgt/VISIBILITY_INLINES_HIDDEN.html)). `-fvisibility-inlines-hidden` also does a little more than what the comment says.

While looking up the flag this uses, I stumbled upon the [GCC visibility guide](https://gcc.gnu.org/wiki/Visibility) and I found their recommendations funnily coincidental :)
> In your build system (Makefile etc), you will probably wish to add the `-fvisibility=hidden` and `-fvisibility-inlines-hidden` options to the command line arguments of every GCC invocation. Remember to test your library thoroughly afterwards, including that all exceptions correctly traverse shared object boundaries.